### PR TITLE
Presentation/key set case insensitive

### PR DESCRIPTION
**Pull request by aurislt7**
Updated KeySet to be case-insensitive

### CODE REVIEW

**Review by grigasp — CHANGES_REQUESTED** [review]

File: presentation/common/src/presentation-common/KeySet.ts [hunk 1]
```diff
@@ -381,7 +391,8 @@ export class KeySet {
   public forEach(callback: (key: Key, index: number) => void) {
     let index = 0;
     this._instanceKeys.forEach((ids: Set<Id64String>, className: string) => {
```
[the remark this answers is not in the document]
Comment (aurislt7) [reply presentation/common/src/presentation-common/KeySet.ts:394]: Added

File: presentation/common/src/presentation-common/KeySet.ts [hunk 2]
```diff
@@ -61,7 +61,8 @@ export interface KeySetJSON {
  */
 export class KeySet {
   // note: all keys are stored as strings because we need ability to find them by value
-  private _instanceKeys: Map<string, Set<string>>; // class name => instance ids
+  private _instanceKeys: Map<string, Set<string>>; // lower case class name => instance ids
+  private _lowerCaseMap: Map<string, string>; // lower case class name => most recent class name
```
[the remark this answers is not in the document]
Comment (aurislt7) [reply presentation/common/src/presentation-common/KeySet.ts:65]: Made improvements

File: common/changes/@bentley/presentation-common/presentation-KeySet_case_insensitive_2020-12-10-10-17.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@bentley/presentation-common",
+      "comment": "Updated KeySet's className to be case-insensitive",
```
[the remark this answers is not in the document]
Comment (aurislt7) [reply common/changes/@bentley/presentation-common/presentation-KeySet_case_insensitive_2020-12-10-10-17.json]: Changed it

File: presentation/common/src/presentation-common/KeySet.ts [hunk 4]
```diff
@@ -165,9 +168,10 @@ export class KeySet {
       if (!set) {
         set = new Set();
         this._instanceKeys.set(entry["0"], set);
+        this._lowerCaseMap.set(entry["0"], (keyset as KeySet)._lowerCaseMap.get(entry["0"])!);
       }
       entry["1"].forEach((id: Id64String) => {
-        if (!pred || pred({ className: entry[0], id }))
+        if (!pred || pred({ className: (keyset as KeySet)._lowerCaseMap.get(entry["0"])!, id }))
```
Comment (grigasp) [inline presentation/common/src/presentation-common/KeySet.ts:174]: It's very inefficient to lookup with the class name on every id. The lookup could be made once at the top of the method.

File: presentation/common/src/presentation-common/KeySet.ts [hunk 5]
```diff
@@ -176,8 +180,11 @@ export class KeySet {
   private addKeySetJSON(keyset: Readonly<KeySetJSON>): void {
     for (const key of keyset.nodeKeys)
       this._nodeKeys.add(JSON.stringify(key));
-    for (const entry of keyset.instanceKeys)
-      this._instanceKeys.set(entry["0"], new Set(entry["1"]));
+    for (const entry of keyset.instanceKeys) {
+      this._instanceKeys.set(entry["0"].toLowerCase(), new Set(entry["1"]));
```
Comment (grigasp) [inline presentation/common/src/presentation-common/KeySet.ts:184]: Could convert the class name to lower case only once

File: presentation/common/src/presentation-common/KeySet.ts [hunk 6]
```diff
@@ -198,9 +205,12 @@ export class KeySet {
...
         this.add({ className: value.classFullName, id: Id64.fromJSON(value.id) } as InstanceKey);
       } else if (Key.isInstanceKey(value)) {
-        if (!this._instanceKeys.has(value.className))
-          this._instanceKeys.set(value.className, new Set());
-        this._instanceKeys.get(value.className)!.add(value.id);
+        if (!this._instanceKeys.has(value.className.toLowerCase())) {
+          this._instanceKeys.set(value.className.toLowerCase(), new Set());
+          this._lowerCaseMap.set(value.className.toLowerCase(), value.className);
+        }
+        this._lowerCaseMap.set(value.className.toLowerCase(), value.className);
```
Comment (grigasp) [inline presentation/common/src/presentation-common/KeySet.ts:212]: Could convert the class name to lower case only once.

**Review by grigasp — CHANGES_REQUESTED** [review]

File: presentation/common/src/presentation-common/KeySet.ts (hunk 6 above)
Comment (aurislt7) [reply presentation/common/src/presentation-common/KeySet.ts]: Fixed it

File: presentation/common/src/presentation-common/KeySet.ts (hunk 5 above)
Comment (aurislt7) [reply presentation/common/src/presentation-common/KeySet.ts]: Fixed it

File: presentation/common/src/presentation-common/KeySet.ts (hunk 4 above)
Comment (aurislt7) [reply presentation/common/src/presentation-common/KeySet.ts]: Yeeh, fixed it

**Review by grigasp — APPROVED** [review]

**Pull request merged**